src/demux/readers/eos/MongoActionReader.ts

```diff
@@ -0,0 +1,51 @@
+import { AbstractActionReader } from "../AbstractActionReader"
+import { MongoBlock } from "./MongoBlock"
+
+import { MongoClient } from "mongodb"
+
+/**
+ * Implementation of an ActionReader that polls a mongodb.
+ */
+export class MongoActionReader extends AbstractActionReader {
+  protected mongoEndpoint: string
+  protected mongodb: any
+  constructor(
+    mongoEndpoint: string = "mongodb://127.0.0.1:27017",
+    public startAtBlock: number = 1,
+    protected onlyIrreversible: boolean = false,
+    protected maxHistoryLength: number = 600,
+  ) {
+    super(startAtBlock, onlyIrreversible, maxHistoryLength)
+    this.mongoEndpoint = mongoEndpoint
+    this.mongodb = null
+  }
+
+  public async initialize() {
+    const mongoInstance = await MongoClient.connect(this.mongoEndpoint, { useNewUrlParser: true })
+    this.mongodb = await mongoInstance.db("EOS")
+  }
+
+  public async getHeadBlockNumber(): Promise<number> {
```

Comment:
per my comment in the example.  We should add a common method "throwIfNotInitialized" that will throw an error if the mongodb instance isn't initialized,  that method can just check initialized === true and throw the error.  Adding that method to the top of all methods needing initialization will handle it for us.  Let me know if you want to discuss with me.